### PR TITLE
Fix for non closure of input stream for Copy utility

### PR DESCRIPTION
**Pull request by sagarbhavsar4328**
Fix for Issues #861 and #908
Closing the exec process (or WebSocket) post receiving the exit message (i.e. when stream == 3).

### CODE REVIEW

File: util/src/main/java/io/kubernetes/client/Copy.java [hunk 1]
```diff
@@ -150,14 +150,15 @@ public void copyDirectoryFromPod(
         }
       }
     }
-    try {
+    // No need to call waitFor() as proc is closed on exit(stream == 3).
+    /*try {
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Copy.java:154]: Please just delete code instead of commenting it out.

**Comment by brendandburns** [comment]
One minor comment, otherwise LGTM.

File: util/src/main/java/io/kubernetes/client/Copy.java (hunk 1 above)
Comment (sagarbhavsar4328) [reply util/src/main/java/io/kubernetes/client/Copy.java]: review comments are incorporated.

File: util/src/main/java/io/kubernetes/client/Exec.java [hunk 2]
```diff
@@ -371,6 +371,8 @@ protected void handleMessage(int stream, InputStream inStream) throws IOExceptio
                   }
                 }
                 inStream.close();
+                // The websocket needs to be closed.
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/Exec.java:374]:
```suggestion
                // Stream ID of `3` delivers exec errors from kubelet and it means the exec is somehow working abnormally, the websocket connection shall be closed by itself upon these errors.
```

File: examples/src/main/java/io/kubernetes/client/examples/ExecExample.java [hunk 3]
```diff
@@ -98,6 +98,19 @@ public void run() {
             });
     out.start();
 
+    Thread error =
```
Comment (yue9944882) [inline examples/src/main/java/io/kubernetes/client/examples/ExecExample.java:101]: this looks irrelevant w/ the purpose of this pull? but the change is good

File: util/src/main/java/io/kubernetes/client/Copy.java [hunk 4]
```diff
@@ -150,14 +150,6 @@ public void copyDirectoryFromPod(
         }
       }
     }
-    try {
-      int status = proc.waitFor();
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/Copy.java:154]: what's the point of deleting this? the method is always supposed to be hanging until the copy completes, isn't it?

File: util/src/main/java/io/kubernetes/client/Exec.java [hunk 5]
```diff
@@ -371,6 +371,8 @@ protected void handleMessage(int stream, InputStream inStream) throws IOExceptio
...
+                this.close();
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/Exec.java:375]: and we should call `ExecProcess.this.notifyAll();` in the end, are we?

File: util/src/main/java/io/kubernetes/client/Copy.java (hunk 4 above)
Comment (sagarbhavsar4328) [reply util/src/main/java/io/kubernetes/client/Copy.java:154]: When the response from k8s API (exit call) is lightning fast, `ExecProcess.this.notifyAll();` gets called before `proc.waitFor();` and this method gets hanged for an infinite duration as no one is left to call `notify()` or `notifyAll()` method on this `proc` object.

File: util/src/main/java/io/kubernetes/client/Exec.java (hunk 5 above)
Comment (sagarbhavsar4328) [reply util/src/main/java/io/kubernetes/client/Exec.java:375]: yes, we are but that will just release the locks and resume the blocked threads (if any) but it won't close the WebSocket.
I believe the WebSocket connection is not getting close by itself nor the input stream.
And eventually, the code gets hanged at `ByteStreams.copy(dataStream, System.out);` in `CopyExample`.

File: examples/src/main/java/io/kubernetes/client/examples/ExecExample.java (hunk 3 above)
Comment (sagarbhavsar4328) [reply examples/src/main/java/io/kubernetes/client/examples/ExecExample.java:101]: I will move this to a different PR.

File: util/src/main/java/io/kubernetes/client/Exec.java (hunk 5 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/Exec.java:375]: > yes, we are but that will just release the locks and resume the blocked threads (if any)

and it will improve the robustness. how about releasing a CountDownLatch here?

> I believe the WebSocket connection is not getting close by itself nor the input stream.
And eventually, the code gets hanged at ByteStreams.copy(dataStream, System.out); in CopyExample.

if i'm understanding correctly, as the websocket is a full-duplex communication, there's supposed to be two "close" action for completely shutting down the connection:
- one is at the server-side, triggerred by the k8s apiserver.
- the other is at the client-side, by invoking `socket.close()` somewhere else

so the websocket isn't "closed" as long as either side isn't yet closed. after the k8s apiserver signals an exit code by the pipe id-3 (reaching this line), it should do the server-side close, which on the other hand triggers the client-side close. so theoretically we don't need to close it again upon exit codes. but as the close call looks idempotent, closing it twice wouldn't be risky..

Comment (sagarbhavsar4328) [reply util/src/main/java/io/kubernetes/client/Exec.java:375]: > after the k8s apiserver signals an exit code by the pipe id-3 (reaching this line), it should do the server-side close,

Unfortunately, at this point, k8s apiserver is NOT closing the connection from there end but on the other side, if the client itself initiates the WebSocket close, post that k8s apiserver also invokes `close()` method of `WebSocketStreamHandler`.

File: util/src/main/java/io/kubernetes/client/Copy.java (hunk 4 above)
Comment (sagarbhavsar4328) [reply util/src/main/java/io/kubernetes/client/Copy.java:154]: > m how do you think about replacing the `wait/notify` w/ `CountDownLatch`? 

`CountDownLatch` seems to be a good idea, let me try it out.

**Comment by brendandburns** [comment]
Closing this in favor of #933
